xds/src/test/java/io/grpc/xds/sds/ClientSslContextProviderFactoryTest.java

```diff
@@ -0,0 +1,159 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.sds;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.common.base.Strings;
+import io.envoyproxy.envoy.api.v2.auth.CertificateValidationContext;
+import io.envoyproxy.envoy.api.v2.auth.CommonTlsContext;
+import io.envoyproxy.envoy.api.v2.auth.SdsSecretConfig;
+import io.envoyproxy.envoy.api.v2.auth.TlsCertificate;
+import io.envoyproxy.envoy.api.v2.auth.UpstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource.ApiType;
+import io.envoyproxy.envoy.api.v2.core.ConfigSource;
+import io.envoyproxy.envoy.api.v2.core.DataSource;
+import io.envoyproxy.envoy.api.v2.core.GrpcService;
+import io.envoyproxy.envoy.api.v2.core.GrpcService.GoogleGrpc;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link ClientSslContextProviderFactory}. */
+@RunWith(JUnit4.class)
+public class ClientSslContextProviderFactoryTest {
+
+  private static final String CLIENT_PEM_FILE = "client.pem";
+  private static final String CLIENT_KEY_FILE = "client.key";
+  private static final String CA_PEM_FILE = "ca.pem";
+
+  ClientSslContextProviderFactory clientSslContextProviderFactory =
+      new ClientSslContextProviderFactory();
+
+  static CommonTlsContext buildCommonTlsContextFromSdsConfigForTlsCertificate(
+      String name, String targetUri, String trustCa) {
+
+    CommonTlsContext.Builder builder =
+        CommonTlsContext.newBuilder()
+            .addTlsCertificateSdsSecretConfigs(
+                SdsSecretConfig.newBuilder()
+                    .setName(name)
+                    .setSdsConfig(
+                        ConfigSource.newBuilder()
+                            .setApiConfigSource(
+                                ApiConfigSource.newBuilder()
+                                    .setApiType(ApiType.GRPC)
+                                    .addGrpcServices(
+                                        GrpcService.newBuilder()
+                                            .setGoogleGrpc(
+                                                GoogleGrpc.newBuilder()
+                                                    .setTargetUri(targetUri)
+                                                    .build())
+                                            .build())
+                                    .build())
+                            .build())
+                    .build());
```

Comment:
nit: this is my personal preference, so you can ignore. but i think this is abit ugly and you can flatten it.

Reply:
ok